### DISCUSSION TITLE
Time series simulation results recycling

### DISCUSSION THREAD

**Discussion by prautio-jyu** [e2nIEE/pandapipes#646] [General]
Hi all, especially pandapipes developers,

I've been working on a district heating network model based on real-world data. A question arose when I started with the time series simulation: What's the situation on simulation results recycling for pandapipes? I see that this feature is implemented for pandapower, but disabled for pandapipes. Furthermore, are there any expected simulation issues if I implement simulation results recycling myself as a iterative write-read loop?

**Comment by dlohmeier** [comment]
Hi @prautio-jyu,
we do not have this feature implemented in pandapipes. There is a special option called "reuse_internal_data", but I am currently not sure if it works for district heating simulations with the changes that we made to the connectivity check. Our recent focus has not been on internal performance, but rather stability wrt. district heating simulation and some first steps towards transient simulations. Do you need to consider transient effects? One main caveat when trying to recycle results is that the hydraulics and heat transfer calculation have different internal structures identified by different connectivity checks. Currently, the internal structure is overridden between the two calculation modes, making it harder to save results for each of them to the next time step. Would you like to implement your own code into the pipeflow or what are you thinking of?

**Comment by prautio-jyu** [comment]
Hi @dlohmeier,

Thanks for your reply. This clarifies things a lot. I'm working with this as a junior programmer/research assistant and unfortunately it's unlikely that I can find the resources to achieve a distribution ready transient simulation solution. For the time being I'm considering doing a bit of a interim hack job wherein I'd build a for-loop to recycle the simulation results as input parameters for each transient simulation time step. This solution would sidestep the current time series functionality, and in essence my inquiry was about whether or not this would produce a satisfactory simulation. I'm probably going forward with it in any case, if only as a nice experiment.

By the way, another thing I'm working on is a simulation results explorer using Bokeh and HoloViz Python packages. I'll have to see if I can release this into the public when it's in a more presentable state.

**Reply by dlohmeier** [reply]
Hi @prautio-jyu ,
we have been working on making transient simulations work with pandapipes for a while now, but this is still work in progress, which is why it is not included in the develop branch yet. If you like, you can have a look at [dlohmeier/tht_heat_consumer](https://github.com/dlohmeier/pandapipes/tree/tht_heat_consumer) and try if the transient simulation with "transient=True" works for your case. If you encounter any problems, you can let me know. One known issue that we haven't overvome yet is dealing with no-flow branches inside the solver, as we need to model a sensible cool-down there.

**Reply by prautio-jyu** [reply]
Hello @dlohmeier,

Thanks again! As it happens I'm already using the development branch and have modeled sinks as heat consumers, so I'll try the transient simulation function you suggested. We'll have to see how well the simulation results coincide with real world measurements; I think I'll be able to at least post some aggregated statistics about this, but I estimate data gathering and analysis will take a couple of months after simulation development is done and hinges on whether or not we can secure funding to continue working. I'll know more about this in the near future.
